File: sync/src/main/scala/com/fan/mysql/parser/FormatDescriptionEventParser.java
package com.fan.mysql.parser;


import com.fan.mysql.binlog.BinlogEventParser;
import com.fan.mysql.dbsync.BinlogContext;
import com.fan.mysql.dbsync.LogBuffer;
import com.fan.mysql.event.BinlogEvent;
import com.fan.mysql.event.EventHeader;
import com.fan.mysql.event.impl.FormatDescriptionEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/*
 * ---------------------------------------------
 * 2                binlog-version
 * string[50]       mysql-server version
 * 4                create timestamp
 * 1                event header length
 * string[p]        event type header lengths
 * ---------------------------------------------
 * */
public class FormatDescriptionEventParser implements BinlogEventParser {

	private static final Logger logger = LoggerFactory.getLogger(FormatDescriptionEventParser.class);

	public static final int LOG_EVENT_HEADER_LEN = 19;

	public static final int BINLOG_VER_OFFSET = 0;
	public static final int SERVER_VER_OFFSET = 2;
	public static final int SERVER_VER_LEN = 50;
	public static final int CREATE_TIMESTAMP_OFFSET = SERVER_VER_OFFSET + SERVER_VER_LEN;
	public static final int CREATE_TIMESTAMP_LEN = 4;
	public static final int EVENT_HEADER_LEN_OFFSET = (CREATE_TIMESTAMP_OFFSET + CREATE_TIMESTAMP_LEN);

	public static final int[] checksumVersionSplit = { 5, 6, 1 };
	public static final long checksumVersionProduct = (checksumVersionSplit[0] * 256 + checksumVersionSplit[1]) * 256
			+ checksumVersionSplit[2];

	public FormatDescriptionEventParser() {
	}

	public BinlogEvent parse(LogBuffer buffer, EventHeader header, BinlogContext context) {
		FormatDescriptionEvent event = new FormatDescriptionEvent(header);
		final int eventPos = buffer.position();
		// read version
		buffer.position(eventPos);
		int binlogVersion = buffer.getUint16();
		String serverVersion = buffer.getFixString(SERVER_VER_LEN);
		// read create time
		buffer.position(eventPos - LOG_EVENT_HEADER_LEN + CREATE_TIMESTAMP_OFFSET);
		long createTime = buffer.getUint32();
		// read event header length
		buffer.position(eventPos + EVENT_HEADER_LEN_OFFSET);
		int eventHeaderLen = buffer.getUint8();
		if (eventHeaderLen < LOG_EVENT_HEADER_LEN) {
			logger.error("common header length must >=" + LOG_EVENT_HEADER_LEN);
		}
		int numberOfEventTypes = buffer.limit() - (eventPos + EVENT_HEADER_LEN_OFFSET + 1);
		short[] postHeaderLen = new short[numberOfEventTypes];
		for (int i = 0; i < numberOfEventTypes; i++) {
			postHeaderLen[i] = (short) buffer.getUint8();
		}
		event.setBinlogVersion(binlogVersion);
		event.setServerVersion(serverVersion);
		event.setCreateTime(createTime);
		event.setCommonHeaderLen(eventHeaderLen);
		event.setPostHeaderLen(postHeaderLen);
		context.setFormatDescription(event);
		return event;
	}

	public static long versionProduct(int[] versionSplit) {
		return ((versionSplit[0] * 256 + versionSplit[1]) * 256 + versionSplit[2]);
	}

	public static void doServerVersionSplit(String serverVersion, int[] versionSplit) {
		String[] split = serverVersion.split("\\.");
		if (split.length < 3) {
			versionSplit[0] = 0;
			versionSplit[1] = 0;
			versionSplit[2] = 0;
		} else {
			int j = 0;
			for (int i = 0; i <= 2; i++) {
				String str = split[i];
				for (j = 0; j < str.length(); j++) {
					if (!Character.isDigit(str.charAt(j))) {
						break;
					}
				}
				if (j > 0) {
					versionSplit[i] = Integer.valueOf(str.substring(0, j), 10);
				} else {
					versionSplit[0] = 0;
					versionSplit[1] = 0;
					versionSplit[2] = 0;
				}
			}
		}
	}

}

File: async/src/main/scala/com/fan/mysql/async/binlog/parse/FormatDescriptionEventParser.scala
package com.fan.mysql.async.binlog.parse

import com.fan.mysql.async.binlog.event.impl.FormatDescriptionEvent
import com.fan.mysql.async.binlog.event.{BinlogEvent, EventHeader}
import com.fan.mysql.async.binlog.{BinlogDumpContext, BinlogEventParser}
import com.fan.mysql.async.util.ChannelWrapper._
import com.fan.mysql.async.util.Log.Logging
import io.netty.buffer.ByteBuf

object FormatDescriptionEventParser {
  final val LOG_EVENT_HEADER_LEN = 19

  final val BINLOG_VER_OFFSET = 0
  final val SERVER_VER_OFFSET = 2
  final val SERVER_VER_LEN = 50
  final val CREATE_TIMESTAMP_OFFSET: Int = SERVER_VER_OFFSET + SERVER_VER_LEN
  final val CREATE_TIMESTAMP_LEN = 4
  final val EVENT_HEADER_LEN_OFFSET: Int = CREATE_TIMESTAMP_OFFSET + CREATE_TIMESTAMP_LEN

  final val checksumVersionSplit: Array[Int] = Array(5, 6, 1)
  final val checksumVersionProduct
    : Long = (checksumVersionSplit(0) * 256 + checksumVersionSplit(1)) * 256 + checksumVersionSplit(
    2)

  def doServerVersionSplit(serverVersion: String, versionSplit: Array[Int]): Unit = {
    val split = serverVersion.split("\\.")
    if (split.length < 3) {
      versionSplit(0) = 0
      versionSplit(1) = 0
      versionSplit(2) = 0
    } else {
      var j = 0
      for (i <- 0 to 2) {
        val str = split(i)
        j = 0
        import scala.util.control.Breaks._
        breakable {
          while (j < str.length) {
            if (!Character.isDigit(str.charAt(j)))
              break
            j += 1
          }
        }
        if (j > 0) versionSplit(i) = Integer.valueOf(str.substring(0, j), 10)
        else {
          versionSplit(0) = 0
          versionSplit(1) = 0
          versionSplit(2) = 0
        }
      }
    }
  }

  def versionProduct(versionSplit: Array[Int]): Long =
    (versionSplit(0) * 256 + versionSplit(1)) * 256 + versionSplit(2)
}

class FormatDescriptionEventParser extends BinlogEventParser with Logging {

  import FormatDescriptionEventParser._

  override def parse(buffer: ByteBuf,
                     header: EventHeader,
                     context: BinlogDumpContext): BinlogEvent = {
    val event = new FormatDescriptionEvent(header)

    val eventPos = buffer.readerIndex()

    // read version
    val binlogVersion = buffer.readUnsignedShort()
    val serverVersion = buffer.readFixedASCIString(SERVER_VER_LEN)

    // read create time
    buffer.readerIndex(eventPos - LOG_EVENT_HEADER_LEN + CREATE_TIMESTAMP_OFFSET)
    val createTime = buffer.readUnsignedInt()

    // read event header length
    buffer.readerIndex(eventPos + EVENT_HEADER_LEN_OFFSET)
    val eventHeaderLen = buffer.readUnsignedByte()
    if (eventHeaderLen < LOG_EVENT_HEADER_LEN)
      logger.error(s"common header length must >= $LOG_EVENT_HEADER_LEN")

    val numberOfEventTypes = buffer.writerIndex() - (eventPos + EVENT_HEADER_LEN_OFFSET + 1)

    val postHeaderLen = new Array[Short](numberOfEventTypes)

    (0 until numberOfEventTypes).foreach(i => postHeaderLen(i) = buffer.readUnsignedByte())

    event.setBinlogVersion(binlogVersion)
    event.setServerVersion(serverVersion)
    event.setCreateTime(createTime)
    event.setCommonHeaderLen(eventHeaderLen)
    event.setPostHeaderLen(postHeaderLen)
    context.setFormatDescription(event)

    event
  }
}
